### PR TITLE
Match MQ-E and CE-E

### PR DESCRIPTION
**Pull request by Yanis42**
I ported the existing changes here, this is a bit ugly but I'm out of ideas at this point lol

european versions are changing a lot (by that I mean more then the north-american versions did), mainly due to the language support, MQ-E adds german and french and CE-E adds spanish and italian (+ the MQ ones obviously), both european versions adds extra files: ``xlText.c`` and ``xlFile.c``, the latter brings symbol oddities with ``xlFileGCN.c`` (some things were moved between these files)

I decided to move the offset tables from rom.c to its own header, it became way too large it's a bit annoying, and the worst function is ``systemSetupGameAll`` because it's a big one and there's lots of changes lol

the only file I didn't match is ``simGCN.c``, I did some changes there but it's very wip (I just don't want to complete it, maybe someday if it's not done until then), otherwise everything else is matched (except libraries/sdk but I'll do that in another PR)

feel free to make suggestions, btw the PR looks big because of the config files, actual code changes aren't that large

### CODE REVIEW

File: configure.py [hunk 1]
```diff
@@ -225,41 +227,43 @@ def MatchingFor(*versions):
...
-            Object(MatchingFor("mq-j", "mq-u", "ce-j", "ce-u"), "emulator/xlList.c"),
-            Object(MatchingFor("mq-j", "mq-u", "ce-j", "ce-u"), "emulator/xlHeap.c"),
-            Object(MatchingFor("mq-j", "mq-u", "ce-j", "ce-u"), "emulator/xlObject.c"),
+            Object(Matching, "emulator/xlCoreGCN.c"),
+            Object(Matching, "emulator/xlPostGCN.c"),
+            Object(Matching, "emulator/xlFileGCN.c"),
+            Object(MatchingFor("mq-e", "ce-e"), "emulator/xlText.c"),
+            Object(Matching, "emulator/xlList.c"),
+            Object(Matching, "emulator/xlHeap.c"),
+            Object(MatchingFor("mq-e", "ce-e"), "emulator/xlFile.c"),
```
Comment (cadmic) [inline configure.py:236]: would this have to stay this way forever? maybe we can mark it as `Matching` by wrapping the entire file in `#if IS_EU`, but I'm sure there's more than one way to do this. Ditto with xlText.c

File: include/emulator/rom.h [hunk 2]
```diff
@@ -77,6 +99,10 @@ typedef struct Rom {
     /* 0x10E70 */ u8 acHeader[64];
     /* 0x10EB0 */ u32* anOffsetBlock;
     /* 0x10EB4 */ s32 nCountOffsetBlocks;
+#if IS_EU
+    /* 0x10EB8 */ RomTagFile tagFile;
+    /* 0x10EC8 */ u32 nChecksum;
+#endif
```
Comment (cadmic) [inline include/emulator/rom.h:105]: more offset comment confusion lol

I'd suggest removing these two offset comments for IS_EU (i.e. offset comments are only for non-EU versions here) or doing the two-offsets-in-one-comment thing we did earlier. What do you think?

File: include/emulator/rom_tables.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    0x01DA6C7F, 0x01DA6C80, 0x01DB675F, 0x01DB6760, 0x01DBD71F, 0x01DBD720, 0x01DC868F, 0x01DC8690, 0x01DCEDCF,
+    0x01DCEDD0, 0x01DD84DF, 0x01DD84E0, 0x01E0D6BF, 0x01E0D6C0, 0x01E1EF6F, 0x01E1EF70, 0x01E2C96F, 0x01E2C970,
+    0x01E36C1F, 0x01E36C20, 0x01E48CFF, 0x01E48D00, 0x01E4F9CF, 0x01E4F9D0, 0x01E75EEF, 0x01E75EF0, 0x01E841DF,
+    0x01E841E0, 0x01E93F2F, 0x01E93F30, 0x01E9EEAF, 0x01E9EEB0, 0x01EAA11F, 0x01EAA120, 0x01EB633F, 0x01EB6340,
+    0x01ECE07F, 0x01ECE080, 0x01ED581F, 0x01ED5820, 0x01EE046F, 0x01EE0470, 0x01EEE44F, 0x01EEE450, 0x01EFABBF,
+    0x01EFABC0, 0x01F0D3DF, 0x01F0D3E0, 0x01F15E7F, 0x01F15E80, 0x01F1F82F, 0x01F1F830, 0x01F30C5F, 0x01F30C60,
+    0x01F4734F, 0x01F47350, 0x01F50B8F, 0x01F50B90, 0x01F594BF, 0x01F594C0, 0x01F63E9F, 0x01F63EA0, 0x01F7531F,
+};
+
+#elif IS_EU
```
Comment (cadmic) [inline include/emulator/rom_tables.h:111]: nit: add newline below

also I think `#elif VERSION == MQ_E || VERSION == CE_E` would be more consistent here

File: src/emulator/rom.c [hunk 4]
```diff
@@ -942,6 +1078,31 @@ bool romGetCode(Rom* pROM, char* acCode) {
...
 
+#if IS_EU
+bool romTestCode(Rom* pROM, char* acCode) {
+    s32 iCode;
+    int nCode1;
+    int nCode2;
+
+    for (iCode = 0; iCode < 4; iCode++) {
+        nCode1 = pROM->acHeader[iCode + 0x3B];
+        if (nCode1 >= 0x61 && nCode1 <= 0x7A) {
```
Comment (cadmic) [inline src/emulator/rom.c:1089]: `nCode1 >= 0x61 && nCode1 <= 0x7A` -> `nCode1 >= 'a' && nCode1 <= 'z'` (turning these uppercase for case-insensitive string compare)
ditto below

File: src/emulator/simGCN.c [hunk 5]
```diff
@@ -2024,7 +2089,10 @@ bool simulatorTestReset(bool IPL, bool forceMenu, bool allowReset, bool usePrevi
 bool simulatorDrawMCardText(void) {
 #if VERSION != MQ_J
     if ((s32)(((TEXPalette*)gpErrorMessageBuffer)->versionNumber) == 0) {
-        xlPostText("Invalid Message Image Data - Assuming SV09", "simGCN.c", 1623);
+        xlPostText("Invalid Message Image Data - Assuming SV09", "simGCN.c",
+                   VERSION == CE_E   ? 1924
+                   : VERSION == MQ_E ? 1836
+                                     : 1623);
```
Comment (cadmic) [inline src/emulator/simGCN.c:2095]: this appears 3 times with the same line numbers, surely this is an inline function. Seems like a big rabbit hole though, maybe add a TODO?

File: src/emulator/xlFile.c [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    NULL,
+    (EventFunc)xlFileEvent,
+};
+#endif
+
+static char* gacValidNumber = "0123456789.xXABCDEFabcdef";
+static char* gacValidSymbol = "~!@#$%^&*()-=_+{}[]|\\:;'<>?,./";
+static char* gacValidLabel = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz0123456789";
+
+char* strchr(const char* str, int chr);
```
Comment (cadmic) [inline src/emulator/xlFile.c:22]: this declaration should go in `string.h`

File: src/emulator/xlFile.c [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+        while (iCharacter < nSizeLine && xlFileGet(pFile, &nCharacter, 1)) {
+            if (nCharacter == '\n') {
+                break;
+            }
+            if (nCharacter == '\r') {
+                xlFileGet(pFile, &nCharacter, 1);
+                if ((nCharacter != '\n') && !xlFileSetPosition(pFile, pFile->nOffset - 1)) {
+                    return false;
+                }
```
Comment (cadmic) [inline src/emulator/xlFile.c:60]:
```suggestion
                if (nCharacter != '\n') {
                    if (!xlFileSetPosition(pFile, pFile->nOffset - 1)) {
                        return false;
                    }
                }
```

File: src/emulator/xlFile.c [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                break;
+            }
+            acLine[iCharacter++] = nCharacter;
+        }
+
+        acLine[iCharacter] = '\0';
+        pFile->nLineNumber++;
+        if ((iCharacter != 0) || (nCharacter == -1)) {
```
Comment (cadmic) [inline src/emulator/xlFile.c:68]: man, all the redundant parentheses in this file hurt my soul. Do you mind cleaning them up? It's kind of a trivial thing though so no worries it not

File: src/emulator/xlFile.c [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        nBase = 0x10;
+        iToken += 2;
+    } else {
+        nBase = 0xA;
+    }
+
+    while (acToken[iToken] != '\0') {
+        nValue *= nBase;
+        if ((acToken[iToken] >= 'A') && (acToken[iToken] <= 'F')) {
+            nValue += acToken[iToken] - '7';
```
Comment (cadmic) [inline src/emulator/xlFile.c:98]: does `acToken[iToken] - 'A' + 10` match? ditto with `acToken[iToken] - 'a' + 10` for the case below

File: src/emulator/xlFileGCN.c [hunk 10]
```diff
@@ -122,17 +130,35 @@ bool xlFileSetPosition(tXL_FILE* pFile, s32 nOffset) {
...
+#endif
             if (!xlHeapTake(&pFile->pBuffer, 0x1024 | 0x30000000)) {
                 return false;
             }
             break;
         case 3:
+#if IS_EU
+            if ((pFile->acLine != NULL) && !xlHeapFree(&pFile->acLine)) {
+                return 0;
+            }
```
Comment (cadmic) [inline src/emulator/xlFileGCN.c:160]:
```suggestion
            if (pFile->acLine != NULL) {
                if (!xlHeapFree(&pFile->acLine)) {
                    return 0;
                }
            }
```

File: src/emulator/xlText.c [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        s32 upper1;
+        s32 upper2;
+
+        iCharacter = nCharacter1;
+        while (iCharacter > 0) {
+            s32 nCharacter1 = *acText1;
+            s32 nCharacter2 = *acText2;
+
+            upper1 = nCharacter1;
+            if ((nCharacter1 >= 0x61) && (upper1 <= 0x7A)) {
```
Comment (cadmic) [inline src/emulator/xlText.c:40]: `nCharacter1 >= 'a' && upper1 <= 'z'` (ditto below)

**Comment by cadmic** [comment]
I'm not sure about incbin for rom arrays tbh, it makes it harder to actually view the data, and as you pointed out the type is wrong. Maybe it doesn't belong in this PR either way though

File: include/emulator/rom.h (hunk 2 above)
Comment (Yanis42) [reply include/emulator/rom.h:105]: I was thinking of this:
```c
typedef struct Rom {
    /*  JP/US     EU   */
    /* 0x00000 0x00000 */ void* pHost;
    /* 0x00004 0x00004 */ void* pBuffer;
    /* 0x00008 0x00008 */ bool bFlip;
    /* 0x0000C 0x0000C */ bool bLoad;
    /* 0x00010 0x00010 */ char acNameFile[513];
    /* 0x00214 0x00214 */ u32 nSize;
    /* 0x00218 0x00218 */ RomModeLoad eModeLoad;
    /* 0x0021C 0x0021C */ RomBlock aBlock[4096];
    /* 0x1021C 0x1021C */ u32 nTick;
    /* 0x10220 0x10220 */ u8* pCacheRAM;
    /* 0x10224 0x10224 */ u8 anBlockCachedRAM[1024]; // Bitfield, one bit per block
    /* 0x10624 0x10624 */ u8 anBlockCachedARAM[2046]; // Bitfield, one bit per block
    /* 0x10E24 0x10E24 */ RomCopyState copy;
    /* 0x10E38 0x10E38 */ RomLoadState load;
    /* 0x10E68 0x10E68 */ s32 nCountBlockRAM;
    /* 0x10E6C 0x10E6C */ s32 nSizeCacheRAM;
    /* 0x10E70 0x10E70 */ u8 acHeader[64];
    /* 0x10EB0 0x10EB0 */ u32* anOffsetBlock;
    /* 0x10EB4 0x10EB4 */ s32 nCountOffsetBlocks;
#if IS_EU
    /*   N/A   0x10EB8 */ RomTagFile tagFile;
    /*   N/A   0x10EC8 */ u32 nChecksum;
#endif
    /* 0x10EB8 0x10ECC */ DVDFileInfo fileInfo;
    /* 0x10EF4 0x10F08 */ s32 offsetToRom;
} Rom; // size = 0x10EF8 (JP/US), 0x10F0C (EU)
```

File: include/emulator/rom_tables.h (hunk 3 above)
Comment (Yanis42) [reply include/emulator/rom_tables.h]: I did the reverse thing and used ``IS_JP``/``IS_US`` instead (also to address the last comment I restored this file)

File: src/emulator/xlFile.c (hunk 8 above)
Comment (Yanis42) [reply src/emulator/xlFile.c:68]: I was different back then :Oldge:

**Review by cadmic — APPROVED** [review]
apologies for forgetting about this

**Pull request merged**